File: sample-videochat-conference-java/src/main/java/com/quickblox/sample/videochat/conference/java/fragments/VideoConversationFragment.java
package com.quickblox.sample.videochat.conference.java.fragments;

import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.ToggleButton;

import com.quickblox.conference.ConferenceSession;
import com.quickblox.conference.view.QBConferenceSurfaceView;
import com.quickblox.sample.videochat.conference.java.R;
import com.quickblox.sample.videochat.conference.java.adapters.OpponentsFromCallAdapter;
import com.quickblox.sample.videochat.conference.java.utils.ToastUtils;
import com.quickblox.videochat.webrtc.callbacks.QBRTCSessionStateCallback;
import com.quickblox.videochat.webrtc.view.QBRTCVideoTrack;

import org.webrtc.CameraVideoCapturer;

import java.io.Serializable;

/**
 * QuickBlox team
 */
public class VideoConversationFragment extends BaseConversationFragment implements Serializable,
        QBRTCSessionStateCallback<ConferenceSession>, OpponentsFromCallAdapter.OnAdapterEventListener {

    private String TAG = VideoConversationFragment.class.getSimpleName();

    private ToggleButton cameraToggle;
    private CameraState cameraState = CameraState.DISABLED_FROM_USER;

    private QBRTCVideoTrack localVideoTrack;
    protected boolean isCurrentCameraFront;


    @Override
    protected void initViews(View view) {
        isCurrentCameraFront = true;
        cameraToggle = (ToggleButton) view.findViewById(R.id.toggle_camera);
        cameraToggle.setVisibility(View.VISIBLE);
        super.initViews(view);
    }

    @Override
    protected void actionButtonsEnabled(boolean inability) {
        super.actionButtonsEnabled(inability);
        cameraToggle.setEnabled(inability);
        // inactivate toggle buttons
        cameraToggle.setActivated(inability);
    }

    @Override
    public void onStart() {
        super.onStart();
        Log.i(TAG, "onStart");
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.d(TAG, "onResume");
        // If user changed camera state few times and last state was CameraState.ENABLED_FROM_USER
        // than we turn on cam, else we nothing change
        if (cameraState != VideoConversationFragment.CameraState.DISABLED_FROM_USER) {
            cameraToggle.setChecked(true);
            toggleCamera(true);
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.i(TAG, "onCreate");
    }

    @Override
    public void onPause() {
        // If camera state is CameraState.ENABLED_FROM_USER or CameraState.NONE
        // than we turn off cam
        if (cameraState != CameraState.DISABLED_FROM_USER) {
            toggleCamera(false);
        }
        super.onPause();
    }

    @Override
    public void onStop() {
        super.onStop();
        Log.d(TAG, "onStop");
    }


    @Override
    public void onDetach() {
        super.onDetach();
        Log.d(TAG, "onDetach");
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Log.d(TAG, "onDestroyView");
    }

    @Override
    protected void setActionButtonsInvisible() {
        super.setActionButtonsInvisible();
        cameraToggle.setVisibility(View.INVISIBLE);
    }


    protected void initButtonsListener() {
        super.initButtonsListener();

        cameraToggle.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (cameraState != CameraState.DISABLED_FROM_USER) {
                    toggleCamera(isChecked);
                }
            }
        });
    }

    private void switchCamera(final MenuItem item) {
        if (cameraState == CameraState.DISABLED_FROM_USER) {
            return;
        }
        cameraToggle.setEnabled(false);
        conversationFragmentCallbackListener.onSwitchCamera(new CameraVideoCapturer.CameraSwitchHandler() {
            @Override
            public void onCameraSwitchDone(boolean b) {
                Log.d(TAG, "camera switched, bool = " + b);
                isCurrentCameraFront = b;
                updateSwitchCameraIcon(item);
                toggleCameraInternal();
            }

            @Override
            public void onCameraSwitchError(String s) {
                Log.d(TAG, "camera switch error " + s);
                ToastUtils.shortToast(getString(R.string.camera_swicth_failed) + s);
                cameraToggle.setEnabled(true);
            }
        });
    }

    private void updateSwitchCameraIcon(final MenuItem item) {
        if (isCurrentCameraFront) {
            Log.d(TAG, "CameraFront now!");
            item.setIcon(R.drawable.ic_camera_front);
        } else {
            Log.d(TAG, "CameraRear now!");
            item.setIcon(R.drawable.ic_camera_rear);
        }
    }

    private void toggleCameraInternal() {
        Log.d(TAG, "Camera was switched!");
        updateVideoView(localVideoView, isCurrentCameraFront);
        toggleCamera(true);
    }

    private void toggleCamera(boolean isNeedEnableCam) {
        if (currentSession != null && currentSession.getMediaStreamManager() != null) {
            conversationFragmentCallbackListener.onSetVideoEnabled(isNeedEnableCam);
        }
        if (!cameraToggle.isEnabled()) {
            cameraToggle.setEnabled(true);
        }
    }

    @Override
    protected void fillVideoView(QBConferenceSurfaceView videoView, QBRTCVideoTrack videoTrack,
                                 boolean remoteRenderer) {
        super.fillVideoView(videoView, videoTrack, remoteRenderer);
        if (!remoteRenderer) {
            updateVideoView(videoView, isCurrentCameraFront);
        }
    }

    ////////////////////////////  callbacks from QBRTCClientVideoTracksCallbacks ///////////////////
    @Override
    public void onLocalVideoTrackReceive(ConferenceSession qbrtcSession, final QBRTCVideoTrack videoTrack) {
        Log.d(TAG, "onLocalVideoTrackReceive() run");
        localVideoTrack = videoTrack;
        cameraState = CameraState.NONE;
        actionButtonsEnabled(true);

        if (localVideoView != null) {
            fillVideoView(localVideoView, localVideoTrack, false);
        }
    }

    /////////////////////////////////////////    end    ////////////////////////////////////////////

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.camera_switch:
                Log.d(TAG, "camera_switch");
                switchCamera(item);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private enum CameraState {
        NONE,
        DISABLED_FROM_USER,
        ENABLED_FROM_USER
    }
}

File: sample-videochat-conference-kotlin/src/main/java/com/quickblox/sample/videochat/conference/kotlin/fragments/VideoConversationFragment.kt
package com.quickblox.sample.videochat.conference.kotlin.fragments

import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.ToggleButton
import com.quickblox.conference.ConferenceSession
import com.quickblox.conference.view.QBConferenceSurfaceView
import com.quickblox.sample.videochat.conference.kotlin.R
import com.quickblox.sample.videochat.conference.kotlin.adapter.OpponentsFromCallAdapter
import com.quickblox.sample.videochat.conference.kotlin.utils.shortToast
import com.quickblox.videochat.webrtc.callbacks.QBRTCSessionStateCallback
import com.quickblox.videochat.webrtc.view.QBRTCVideoTrack
import org.webrtc.CameraVideoCapturer
import org.webrtc.SurfaceViewRenderer
import java.io.Serializable


open class VideoConversationFragment : BaseConversationFragment(), Serializable,
        QBRTCSessionStateCallback<ConferenceSession>, OpponentsFromCallAdapter.OnAdapterEventListener {

    private val TAG = VideoConversationFragment::class.java.simpleName

    private enum class CameraState {
        NONE,
        DISABLED_FROM_USER,
        ENABLED_FROM_USER
    }

    private lateinit var cameraToggle: ToggleButton

    private var cameraState = CameraState.DISABLED_FROM_USER
    private var localVideoTrack: QBRTCVideoTrack? = null

    protected var isCurrentCameraFront: Boolean = false

    override fun initViews(view: View) {
        isCurrentCameraFront = true
        cameraToggle = view.findViewById<View>(R.id.toggle_camera) as ToggleButton
        cameraToggle.visibility = View.VISIBLE
        super.initViews(view)
    }

    override fun actionButtonsEnabled(inability: Boolean) {
        super.actionButtonsEnabled(inability)
        cameraToggle.isEnabled = inability
        // inactivate toggle buttons
        cameraToggle.isActivated = inability
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume")
        // If user changed camera state few times and last state was CameraState.ENABLED_FROM_USER
        // than we turn on cam, else we nothing change
        if (cameraState != VideoConversationFragment.CameraState.DISABLED_FROM_USER) {
            cameraToggle.isChecked = true
            toggleCamera(true)
        }
    }

    override fun onPause() {
        // If camera state is CameraState.ENABLED_FROM_USER or CameraState.NONE
        // than we turn off cam
        if (cameraState != CameraState.DISABLED_FROM_USER) {
            toggleCamera(false)
        }
        super.onPause()
    }

    override fun setActionButtonsInvisible() {
        super.setActionButtonsInvisible()
        cameraToggle.visibility = View.INVISIBLE
    }

    override fun initButtonsListener() {
        super.initButtonsListener()

        cameraToggle.setOnCheckedChangeListener { buttonView, isChecked ->
            if (cameraState != CameraState.DISABLED_FROM_USER) {
                toggleCamera(isChecked)
            }
        }
    }

    private fun switchCamera(item: MenuItem) {
        if (cameraState == CameraState.DISABLED_FROM_USER) {
            return
        }
        cameraToggle.isEnabled = false
        conversationFragmentCallbackListener?.onSwitchCamera(object : CameraVideoCapturer.CameraSwitchHandler {
            override fun onCameraSwitchDone(frontCamera: Boolean) {
                Log.d(TAG, "camera switched, bool = $frontCamera")
                isCurrentCameraFront = frontCamera
                updateSwitchCameraIcon(item)
                toggleCameraInternal()
            }

            override fun onCameraSwitchError(message: String) {
                Log.d(TAG, "camera switch error $message")
                shortToast(getString(R.string.camera_swicth_failed) + message)
                cameraToggle.isEnabled = true
            }
        })
    }

    private fun updateSwitchCameraIcon(item: MenuItem) {
        if (isCurrentCameraFront) {
            Log.d(TAG, "CameraFront now!")
            item.setIcon(R.drawable.ic_camera_front)
        } else {
            Log.d(TAG, "CameraRear now!")
            item.setIcon(R.drawable.ic_camera_rear)
        }
    }

    private fun toggleCameraInternal() {
        Log.d(TAG, "Camera was switched!")
        updateVideoView(localVideoView as SurfaceViewRenderer, isCurrentCameraFront)
        toggleCamera(true)
    }

    private fun toggleCamera(isNeedEnableCam: Boolean) {
        if (currentSession != null && currentSession?.mediaStreamManager != null) {
            conversationFragmentCallbackListener?.onSetVideoEnabled(isNeedEnableCam)
        }
        if (!cameraToggle.isEnabled) {
            cameraToggle.isEnabled = true
        }
    }

    override fun fillVideoView(videoView: QBConferenceSurfaceView, videoTrack: QBRTCVideoTrack,
                               remoteRenderer: Boolean) {
        super.fillVideoView(videoView, videoTrack, remoteRenderer)
        if (!remoteRenderer) {
            updateVideoView(videoView, isCurrentCameraFront)
        }
    }

    ////////////////////////////  callbacks from QBRTCClientVideoTracksCallbacks ///////////////////

    override fun onLocalVideoTrackReceive(p0: ConferenceSession?, videoTrack: QBRTCVideoTrack?) {
        Log.d(TAG, "onLocalVideoTrackReceive() run")
        localVideoTrack = videoTrack
        cameraState = CameraState.NONE
        actionButtonsEnabled(true)

        if (localVideoView != null) {
            fillVideoView(localVideoView as QBConferenceSurfaceView, localVideoTrack as QBRTCVideoTrack, false)
        }
    }

    /////////////////////////////////////////    end    ////////////////////////////////////////////

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.camera_switch -> {
                Log.d(TAG, "camera_switch")
                switchCamera(item)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }
}